core/src/main/java/org/apache/calcite/rel/rules/AggregateExpandDistinctAggregatesRule.java

```diff
@@ -366,12 +370,18 @@ private static RelBuilder convertSingletonDistinct(RelBuilder relBuilder,
         final int arg = bottomGroups.size() + nonDistinctAggCallProcessedSoFar;
         final List<Integer> newArgs = ImmutableList.of(arg);
         if (aggCall.getAggregation().getKind() == SqlKind.COUNT) {
+          RelDataTypeFactory typeFactory = aggregate.getCluster().getTypeFactory();
+          RelDataType sumType = typeFactory.getTypeSystem().deriveSumType(typeFactory,
+              aggCall.getType());
+          needTopCast |= !sumType.equals(aggCall.getType());
```

Comment:
running new test under debug, i found aggCall.getType() == BIGINT, while sumType == DECIMAL(19, 0) all these representation are identical as i can see, can you explain why we rise needTopCast flag in this case? Additionally i hope we need integration test highliting the bug, ServerTest possibly correct place for it. wdyt ?

Reply:
Because the transformation rule must do an equivalent transformation. So, the type of the output row must be the same as that of the original node. DECIMAL and BIGINT are different types.